org.eclipse.corrosion/src/org/eclipse/corrosion/CorrosionPreferencePage.java

```diff
@@ -382,34 +383,68 @@ private void createCommandPathsPart(Composite container) {
 	private void installCommands() {
 		installButton.setText(Messages.CorrosionPreferencePage_installing);
 		installButton.setEnabled(false);
-		Job.create(Messages.CorrosionPreferencePage_installingRustupCargo, (ICoreRunnable) monitor -> {
-			try {
-				Bundle bundle = CorrosionPlugin.getDefault().getBundle();
-				URL fileURL = FileLocator.toFileURL(bundle.getEntry("scripts/rustup-init.sh")); //$NON-NLS-1$
-				File file = new File(new URI(fileURL.getProtocol(), fileURL.getPath(), null));
-				file.setExecutable(true);
-				String[] command = new String[] { "/bin/bash", "-c", file.getAbsolutePath() + " -y" }; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
-				ProcessBuilder builder = new ProcessBuilder(command);
-				Process process = builder.start();
-				if (process.waitFor() == 0) {
-					CorrosionPreferenceInitializer initializer = new CorrosionPreferenceInitializer();
-					initializer.initializeDefaultPreferences();
-					Display.getDefault().asyncExec(() -> {
-						setInstallRequired(false);
-						performDefaults();
-						setValid(isPageValid());
-					});
-					RustManager.setDefaultToolchain("beta"); //$NON-NLS-1$
-					return;
+		ProgressMonitorDialog dialog = new ProgressMonitorDialog(this.getShell());
+		try {
+			dialog.run(true, true, monitor -> {
+				monitor.beginTask(Messages.CorrosionPreferencePage_installingRustupCargo, 15);
+				try {
+					Bundle bundle = CorrosionPlugin.getDefault().getBundle();
+					URL fileURL = FileLocator.toFileURL(bundle.getEntry("scripts/rustup-init.sh")); //$NON-NLS-1$
+					File file = new File(new URI(fileURL.getProtocol(), fileURL.getPath(), null));
+					file.setExecutable(true);
+					String[] command = new String[] { "/bin/bash", "-c", file.getAbsolutePath() + " -y" }; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+					ProcessBuilder builder = new ProcessBuilder(command);
+					Process process = builder.start();
+					InputStream stream = process.getErrorStream();
+					String newLineString = ""; //$NON-NLS-1$
+					int i;
+					while ((i = stream.read()) != -1 && !monitor.isCanceled()) {
+						char c = (char) i;
+						if (c == '\n') {
+							monitor.subTask(newLineString);
+							monitor.worked(1);
+							newLineString = ""; //$NON-NLS-1$
+						} else {
+							newLineString += c;
+						}
+					}
+					if (monitor.isCanceled()) {
```

Comment:
I believe this should be checks in the while loop. Otherwise cancelling would only happen when stream is closed, which is kind of equivalent to when process is finished.